src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -176,25 +177,60 @@ class DecisionTree(BaseModel):
         if (not leaf_nodes):  # make sure that there are calculated observations
             raise ValueError("The tree has no leaf nodes")
 
-        n_rows = self.n_rows
-        matrix = np.zeros((n_rows, n_rows))
+        ht = {}
         for node in leaf_nodes:
-            indices = node.indices
-            for i in indices:
-                for j in indices:
-                    if scale:
-                        n_node = len(node.indices)
-                        matrix[i, j] += 1/n_node
-                    else:
-                        matrix[i, j] += 1
+            ht[node.id] = node.indices
+        return ht
+
+    def __tree_based_weights(self, hash1, hash2, size_X0, size_X1, scale):
+        matrix = np.zeros((size_X0, size_X1))
+        hash1_keys = hash1.keys()
+        hash2_keys = hash2.keys()
+        for xi in hash1_keys:
+            if xi in hash2_keys:
+                indices_1 = hash1[xi]
+                indices_2 = hash2[xi]
+                if scale == 0: 
+                    val = 1.0/len(indices_1)
+                elif scale == 1:
+                    val = 1.0/(len(indices_1) + len(indices_2))
+                else:
+                    val = 1.0
+                matrix[np.ix_(indices_1, indices_2)] = val
         return matrix
 
-    def predict_leaf_matrix(self, X: np.ndarray|None, scale: bool = False):
+    def similarity(self, X0, X1, scale=True):
+        hash1 = self.predict_leaf(X0)
+        hash2 = self.predict_leaf(X1)
+        if scale:
+            scale = 1
+        else:
+            scale = -1
+        return self.__tree_based_weights(hash1, hash2, X0.shape[0], X1.shape[0],
+                                         scale)
+
+    def predict_tree_based_weights(self, X: np.ndarray|None, scale: bool=True):
```

Comment:
- set default of X to None
- Can we change the name to "tree_based_weights"? Its kind of long with predict and maybe this is not fully necessary...